### PR TITLE
single_match: suggest `if` over `if let` when possible

### PR DESCRIPTION
**Pull request by Jarcho**
fixes: #173
changelog: single_match: suggest `if` over `if let` when possible

### CODE REVIEW

File: clippy_lints/src/matches.rs [hunk 1]
```diff
@@ -717,6 +717,28 @@ fn check_single_match_single_pattern(
...
+        if let PatKind::Ref(pat, _) = pat.kind {
+            peel(pat, count + 1)
+        } else {
+            (pat, count)
+        }
+    }
+    peel(pat, 0)
+}
+
+fn peel_ty_refs(ty: Ty<'_>) -> (Ty<'_>, usize) {
```
Comment (Manishearth) [inline clippy_lints/src/matches.rs:731]: These functions should go in `util` somewhere, and I think they may actually already exist there

File: tests/ui/single_match.rs [hunk 2]
```diff
@@ -81,6 +81,49 @@ fn single_match_know_enum() {
     }
 }
 
+fn issue_173() {
```
Comment (Manishearth) [inline tests/ui/single_match.rs:84]: I'd like some tests that exercise the `&` code

also perhaps we can name this `fn suggest_if` and mention the issue in a comment

**Review by Manishearth — CHANGES_REQUESTED** [review]
Mostly looks good, minor changes

File: clippy_lints/src/matches.rs [hunk 3]
```diff
@@ -728,20 +750,51 @@ fn report_single_match_single_pattern(
...
+        if let Some(trait_id) = cx.tcx.lang_items().structural_peq_trait();
+        if ty.is_integral() || ty.is_char() || ty.is_str() || implements_trait(cx, ty, trait_id, &[]);
+        then {
+            // scrutinee derives PartialEq and the pattern is a constant.
+            let pat_ref_count = match pat.kind {
+                // string literals are already a reference.
+                PatKind::Lit(Expr { kind: ExprKind::Lit(lit), .. }) if lit.node.is_str() => pat_ref_count + 1,
+                _ => pat_ref_count,
+            };
+            let msg = "you seem to be trying to use match for an equality check. Consider using `if`";
```
Comment (giraffate) [inline clippy_lints/src/matches.rs:767]: nits
```suggestion
            let msg = "you seem to be trying to use `match` for an equality check. Consider using `if`";
```

File: clippy_lints/src/matches.rs [hunk 4]
```diff
@@ -728,20 +750,51 @@ fn report_single_match_single_pattern(
...
+                snippet(cx, ex.span, ".."),
+                // PartialEq for different reference counts may not exist.
+                "&".repeat(ty_ref_count - pat_ref_count),
+                snippet(cx, arms[0].pat.span, ".."),
+                expr_block(cx, &arms[0].body, None, "..", Some(expr.span)),
+                els_str,
+            );
+            (msg, sugg)
+        } else {
+            let msg = "you seem to be trying to use match for destructuring a single pattern. Consider using `if let`";
```
Comment (giraffate) [inline clippy_lints/src/matches.rs:779]: nits
```suggestion
            let msg = "you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`";
```

File: clippy_lints/src/matches.rs (hunk 1 above)
Comment (Jarcho) [reply clippy_lints/src/matches.rs]: Both have been moved with a slightly different name to disambiguate between ast, hir, and middle.

File: clippy_lints/src/utils/mod.rs [hunk 5]
```diff
@@ -1668,6 +1668,44 @@ where
...
+        } else {
+            (pat, count)
+        }
+    }
+    peel(pat, 0)
+}
+
+/// Peels off up to the given number of references on the expression. Returns the underlying
+/// expression and the number of references removed.
+pub fn peeln_hir_expr_refs(expr: &'a Expr<'a>, count: usize) -> (&'a Expr<'a>, usize) {
```
Comment (Manishearth) [inline clippy_lints/src/utils/mod.rs:1686]: `peel_n`

**Review by Manishearth — APPROVED** [review]

File: clippy_lints/src/utils/mod.rs (hunk 5 above)
Comment (Jarcho) [reply clippy_lints/src/utils/mod.rs:1686]: I was matching `splitn` from std.

**Pull request merged**